PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx

```diff
@@ -0,0 +1,339 @@
+      registry.add("epReso/hEpResoFT0aTPCneg", "hEpResoFT0aTPCneg; centrality; #Delta#Psi_{sub}", {HistType::kTH2F, {thnAxisCent, thnAxisCosNPhi}});
+      registry.add("epReso/hEpResoFT0mFV0m", "hEpResoFT0mFV0m; centrality; #Delta#Psi_{sub}", {HistType::kTH2F, {thnAxisCent, thnAxisCosNPhi}});
+      registry.add("epReso/hEpResoFT0mTPCpos", "hEpResoFT0mTPCpos; centrality; #Delta#Psi_{sub}", {HistType::kTH2F, {thnAxisCent, thnAxisCosNPhi}});
+      registry.add("epReso/hEpResoFT0mTPCneg", "hEpResoFT0mTPCneg; centrality; #Delta#Psi_{sub}", {HistType::kTH2F, {thnAxisCent, thnAxisCosNPhi}});
+      registry.add("epReso/hEpResoFV0mTPCpos", "hEpResoFV0mTPCpos; centrality; #Delta#Psi_{sub}", {HistType::kTH2F, {thnAxisCent, thnAxisCosNPhi}});
+      registry.add("epReso/hEpResoFV0mTPCneg", "hEpResoFV0mTPCneg; centrality; #Delta#Psi_{sub}", {HistType::kTH2F, {thnAxisCent, thnAxisCosNPhi}});
+    }
+  }; // end init
+
+  /// Compute the Q vector for the candidate's tracks
+  /// \param cand is the candidate
+  /// \param tracksQx is the X component of the Q vector for the tracks
+  /// \param tracksQy is the Y component of the Q vector for the tracks
+  template <typename T1>
+  void GetQvecDtracks(const T1& cand,
+                      std::vector<float>& tracksQx,
+                      std::vector<float>& tracksQy,
+                      float& ampl)
+  {
+    // TODO: add possibility to consider different weights for the tracks, at the only pT is considered;
+    float pXtrack0 = cand.pxProng0();
+    float pYtrack0 = cand.pyProng0();
+    float pTtrack0 = cand.ptProng0();
+    float phiTrack0 = TMath::ATan2(pYtrack0, pXtrack0);
+    float pXtrack1 = cand.pxProng1();
+    float pYtrack1 = cand.pyProng1();
+    float pTtrack1 = cand.ptProng1();
+    float phiTrack1 = TMath::ATan2(pYtrack1, pXtrack1);
+    float pXtrack2 = cand.pxProng2();
+    float pYtrack2 = cand.pyProng2();
+    float pTtrack2 = cand.ptProng2();
+    float phiTrack2 = TMath::ATan2(pYtrack2, pXtrack2);
+
+    tracksQx.push_back(TMath::Cos(harmonic * phiTrack0) * pTtrack0 / ampl);
+    tracksQy.push_back(TMath::Sin(harmonic * phiTrack0) * pTtrack0 / ampl);
+    tracksQx.push_back(TMath::Cos(harmonic * phiTrack1) * pTtrack1 / ampl);
+    tracksQy.push_back(TMath::Sin(harmonic * phiTrack1) * pTtrack1 / ampl);
+    tracksQx.push_back(TMath::Cos(harmonic * phiTrack2) * pTtrack2 / ampl);
+    tracksQy.push_back(TMath::Sin(harmonic * phiTrack2) * pTtrack2 / ampl);
+  }
+
+  /// Compute the delta psi in the range [0, pi/harmonic]
+  /// \param psi1 is the first angle
+  /// \param psi2 is the second angle
+  /// \note Ported from AliAnalysisTaskSECharmHadronvn::GetDeltaPsiSubInRange
+  float GetDeltaPsiInRange(float psi1, float psi2)
+  {
+    float deltaPsi = psi1 - psi2;
+    if (deltaPsi > TMath::Pi() / harmonic) {
+      if (deltaPsi > 0.)
+        deltaPsi -= 2. * TMath::Pi() / harmonic;
+      else
+        deltaPsi += 2. * TMath::Pi() / harmonic;
+    }
+    return deltaPsi;
+  }
```

Comment:
Use `TwoPI` from O2.
This does not look like it can guarantee the returned values to be in the desired range, since it only does one pass of adjustment.
Can this not be replaced with `RecoDecay::constrainAngle` and then divided by `harmonic`?

Reply:
Hi @vkucera ! Thanks a lot for the superquick review and the suggestions. I will implement them. I just have a few doubts. Regarding this comment, if I understood correctly, your idea would be to use the `RecoDecay::constrainAngle`  to subtract/add 2Pi from the DeltaPsi and scale the result by the harmonic. But this is not the same as subtracting/adding a factor of 2Pi/harmonic to the DeltaPsi. Am I right or perhaps I am missing something?

Reply:
I might be misunderstanding of what the function should do but I am suspecting that there are several problems in the current version.

- I assume that `harmonic` is supposed to be a positive integer.
- In that case, if `deltaPsi > TMath::Pi() / harmonic` then also `deltaPsi > 0.` so the `else` clause is never executed and there is no range constrain for `deltaPsi < 0.`.
- I assume that you want to handle all possible values of `deltaPsi`. For the trivial case of `harmonic = 1`, I assume that you want to constrain `deltaPsi` within [0, 2 pi] or [-pi, pi]. Therefore I would expected that for higher harmonics, the range would scale as [0, 2 pi/harmonic] or [-pi/harmonic, pi/harmonic]. So trying to constrain `deltaPsi` to [0, pi/harmonic] does not seem correct to me.

Reply:
You are right that scaling the constrained number would not be correct. The `RecoDecay::constrainAngle` could actually be extended to scale the increments by the harmonic number.

Reply:
I see your point, thanks for the clarification. Indeed, what is missing in line 136 is an absolute value I would say.
( `if (TMath::Abs(deltaPsi) > TMath::Pi() / harmonic) {`). In this way, both conditions are accessed depending on the value of `deltaPsi`.
Regarding the range, I checked the code from [Run2](https://github.com/alisw/AliPhysics/blob/1e829395dcfaf57dde72ae586f068fa631ea1c37/PWGHF/vertexingHF/vHFML/AliAnalysisTaskSECharmHadronvn.cxx#L1088) and indeed the spanned interval is [0, pi/harmonic]. In my understanding, this choice is consistent since the deltaPsi is multiplied by the harmonic (e.g. line 317) before filling the TH1, thus spanning a range of [0, pi].

Reply:
OK, that helps a bit, but still, I would say that for `delta = pi/harmonic + epsilon` the Run 2 function returns `-pi/harmonic + epsilon < 0` which is outside the range you claim to want.

Reply:
So I still suspect that you actually want to constrain the angle to the range [-pi/harmonic, pi/harmonic]. Otherwise, I don't see how you can properly handle negative values.

Reply:
Indeed it is correct to divide by `n` and not by `2` (fortunately for the `v3` in Run 2 we used the SP method, that's also why we did not spot the typo probably).
Regarding the interval of `psi` and `deltaPsi` @fcatalan92 is right, so the implementation of @stefanopolitano is correct.

Reply:
I am still not convinced.
Assuming that |psi| <= pi/2 and the target range is |deltaPsi| <= pi/n (as @fcatalan92 says):
if |psi| <= pi/2, then |psi1 - psi2| <= pi
if (deltaPsi = pi > pi/n), then the function would do: deltaPsi = deltaPsi - 2pi/n = pi(1-2/n)
This deltaPsi will be in the target range if: 1-2/n <= 1/n, which is true only if n <= 3.
So in my opinion, the function is still incorrect, unless the `if` conditions are replaced with `while` loops.

Reply:
Ah, but I guess your point was that the problem is the missing `harmonic` in the calls of `GetEventPlane`, right?

Reply:
Then, indeed `|psi1 - psi2| <= 2pi/n` which works fine.